File: src/main/java/pl/training/workshop/m6/s11_safecomposite/start/Combo.java
package pl.training.workshop.m6.s11_safecomposite.start;

import java.util.ArrayList;
import java.util.List;

import pl.training.workshop.shared.Money;

/** Start: zestaw nadpisuje add() i children(). */
public final class Combo extends MenuComponent {
    private final String name;
    private final List<MenuComponent> children = new ArrayList<>();

    public Combo(String name) {
        this.name = name;
    }

    @Override
    public String name() {
        return name;
    }

    @Override
    public Money price() {
        Money total = Money.ZERO;
        for (MenuComponent child : children) {
            total = total.plus(child.price());
        }
        return total;
    }

    @Override
    public void add(MenuComponent child) {
        children.add(child);
    }

    @Override
    public List<MenuComponent> children() {
        return List.copyOf(children);
    }
}

File: csharp/src/Training.Workshop/M6/S11SafeComposite/Start/Combo.cs
using Training.Workshop.Shared;

namespace Training.Workshop.M6.S11SafeComposite.Start;

/// <summary>Start: zestaw nadpisuje Add() i Children.</summary>
public sealed class Combo : MenuComponent
{
    private readonly string _name;
    private readonly List<MenuComponent> _children = [];

    public Combo(string name)
    {
        _name = name;
    }

    public override string Name => _name;

    public override Money Price
    {
        get
        {
            var total = Money.Zero;
            foreach (var child in _children)
            {
                total = total.Plus(child.Price);
            }
            return total;
        }
    }

    public override void Add(MenuComponent child)
    {
        _children.Add(child);
    }

    public override IReadOnlyList<MenuComponent> Children => _children.ToList().AsReadOnly();
}
